### PR TITLE
Rewrite floating constants placed in read only section

### PR DESCRIPTION
**Pull request by ptomin**
- Prepare unit tests for floating constants placed in read only section

Expected:
```
 "12.75"
```
but was:
```
 "g_dw100000"
```
- Rewrite floating constants placed in read only section.

Compilers can place floating point constant in read only section of
memory. e.g. `constant` can be converted to `[address]`. So we need
rewrite `[address]` back to `constant`.

### CODE REVIEW

File: src/UnitTests/Typing/TypedConstantRewriterTests.cs [hunk 1]
```diff
@@ -321,8 +363,64 @@ public void Tcr_TypelessPointer()
...
-            var e = tcr.Rewrite(c, null, false);
+            var e = RewritePointer(c);
             Assert.AreEqual("&seg0C00->t0200", e.ToString());
         }
+
+        [Test(Description = "Don't rewrite pointer to double constant at read-only memory")]
+        public void Tcr_Real64_ReadOnly_Pointer()
+        {
+            Given_TypedConstantRewriter();
+            Given_UInt64(0x4029800000000000, 0x00100000); // 12.75
```
Comment (uxmal) [inline src/UnitTests/Typing/TypedConstantRewriterTests.cs:374]: Consider renaming this `Given_UInt64_At` to clearly indicate the intent.

**Review by uxmal — COMMENTED** [review]
Looks good; I like the improved code generation. I've added some small nit picks but then this PR is good to go.

File: src/UnitTests/Typing/TypedConstantRewriterTests.cs [hunk 2]
```diff
@@ -97,15 +97,43 @@ private void Given_Segment(ushort selector, string name)
...
+        {
+            return tcr.Rewrite(c, null, false);
+        }
+
+        private Expression RewriteDereferenced(Constant c)
+        {
+            return tcr.Rewrite(c, null, true);
+        }
+
+        [Test]
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/UnitTests/Typing/TypedConstantRewriterTests.cs:165]: Done in c38b69b

File: src/UnitTests/Typing/TypedConstantRewriterTests.cs (hunk 1 above)
Comment (ptomin) [reply src/UnitTests/Typing/TypedConstantRewriterTests.cs]: Done in 344c373

File: src/Decompiler/Typing/TypedConstantRewriter.cs [hunk 3]
```diff
@@ -352,6 +360,28 @@ DataType PromoteToCString(Constant c, DataType charType)
...
+            cReal = null;
+            var pt = dt.ResolveAs<PrimitiveType>();
+            if (pt is null)
+                return false;
+            if (pt.Domain != Domain.Real)
+                return false;
+            // $BUGBUG: ImageReader throws NotImplementedException when
+            // reading of 48-bit real at subjects/Raw/1750A/fff and
+            // subjects/Raw/1750A/trigtst
+            if (pt.BitSize != 32 && pt.BitSize != 64)
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Decompiler/Typing/TypedConstantRewriter.cs:379]: Done in e203a27

**Comment by uxmal** [comment]
Thanks for the PR Pavel

**Pull request merged**